File: backend/services/campaign_manager.py
"""Post-deploy campaign management: edit budgets, add ad sets/ads to live campaigns."""
import json
import logging
from typing import Optional
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select

from models.campaign import Project, AdSet, Ad
from models.creative import Creative
from models.job import DeployLog, DeployStep, DeployStatus
from services.meta_api import MetaAPIService
from services.naming import generate_adset_name, generate_ad_name
from services.utm_builder import build_url_tags

logger = logging.getLogger(__name__)


class CampaignManager:
    def __init__(self, db: AsyncSession, meta_service: MetaAPIService):
        self.db = db
        self.meta = meta_service

    async def update_budget(
        self,
        project_id: int,
        daily_budget: Optional[float] = None,
        lifetime_budget: Optional[float] = None,
    ) -> dict:
        """Update campaign-level budget (CBO)."""
        project = await self._get_project(project_id)
        if not project.meta_campaign_id:
            raise ValueError("Campaign not deployed yet")

        await self.meta.update_campaign_budget(
            project.meta_campaign_id, daily_budget, lifetime_budget
        )

        if daily_budget is not None:
            project.daily_budget = daily_budget
        if lifetime_budget is not None:
            project.lifetime_budget = lifetime_budget
        await self.db.commit()

        return {"status": "updated", "campaign_id": project.meta_campaign_id}

    async def update_adset_budget(
        self, adset_id: int, daily_budget: float
    ) -> dict:
        """Update ad set budget (ABO)."""
        result = await self.db.execute(select(AdSet).where(AdSet.id == adset_id))
        adset = result.scalar_one_or_none()
        if not adset or not adset.meta_adset_id:
            raise ValueError("Ad set not found or not deployed")

        await self.meta.update_adset_budget(adset.meta_adset_id, daily_budget)
        adset.budget = daily_budget
        await self.db.commit()

        return {"status": "updated", "adset_id": adset.meta_adset_id}

    async def add_adset(
        self,
        project_id: int,
        name: str,
        targeting_json: str,
        optimization_goal: str = "OFFSITE_CONVERSIONS",
        bid_strategy: str = "LOWEST_COST_WITHOUT_CAP",
        budget: Optional[float] = None,
        placements_json: Optional[str] = None,
    ) -> AdSet:
        """Add a new ad set to a deployed campaign."""
        project = await self._get_project(project_id)
        if not project.meta_campaign_id:
            raise ValueError("Campaign not deployed yet")

        # Count existing ad sets for variant naming
        existing = await self.db.execute(
            select(AdSet).where(AdSet.project_id == project_id)
        )
        variant_index = len(list(existing.scalars().all()))

        generated_name = generate_adset_name(
            project.name, name, optimization_goal, variant_index
        )

        # Create in Meta
        targeting = json.loads(targeting_json)
        placements = json.loads(placements_json) if placements_json else None

        meta_adset_id = await self.meta.create_adset(
            account_id=project.ad_account_id,
            campaign_id=project.meta_campaign_id,
            name=generated_name,
            optimization_goal=optimization_goal,
            targeting=targeting,
            bid_strategy=bid_strategy,
            daily_budget=budget,
            placements=placements,
        )

        # Save to DB
        adset = AdSet(
            project_id=project_id,
            name=name,
            generated_name=generated_name,
            targeting_json=targeting_json,
            placements_json=placements_json,
            budget=budget,
            bid_strategy=bid_strategy,
            optimization_goal=optimization_goal,
            meta_adset_id=meta_adset_id,
            status="deployed",
        )
        self.db.add(adset)

        # Log
        log = DeployLog(
            project_id=project_id, step=DeployStep.adset,
            entity_name=name, meta_id=meta_adset_id, status=DeployStatus.success,
        )
        self.db.add(log)
        await self.db.commit()

        return adset

    async def add_ad(
        self,
        adset_id: int,
        name: str,
        creative_id: int,
        headline: str = "",
        primary_text: str = "",
        description: str = "",
        cta: str = "SHOP_NOW",
        url: Optional[str] = None,
    ) -> Ad:
        """Add a new ad to an existing ad set."""
        result = await self.db.execute(select(AdSet).where(AdSet.id == adset_id))
        adset = result.scalar_one_or_none()
        if not adset or not adset.meta_adset_id:
            raise ValueError("Ad set not found or not deployed")

        project = await self._get_project(adset.project_id)

        # Get creative info
        creative_result = await self.db.execute(
            select(Creative).where(Creative.id == creative_id)
        )
        creative = creative_result.scalar_one_or_none()
        if not creative:
            raise ValueError(f"Creative {creative_id} not found")

        # Ensure creative is uploaded
        if not creative.meta_image_hash and not creative.meta_video_id:
            # Upload it first
            if creative.media_type and creative.media_type.value == "image":
                creative.meta_image_hash = await self.meta.upload_image(
                    project.ad_account_id, creative.local_path
                )
            else:
                creative.meta_video_id = await self.meta.upload_video(
                    project.ad_account_id, creative.local_path
                )

        # Count existing ads for version naming
        existing_ads = await self.db.execute(
            select(Ad).where(Ad.ad_set_id == adset_id)
        )
        version_index = len(list(existing_ads.scalars().all()))

        generated_name = generate_ad_name(
            adset.generated_name or adset.name,
            creative.base_name or name,
            creative.format.value if creative.format else "",
            version_index,
        )

        url_tags = build_url_tags()
        final_url = url or project.destination_url or ""

        # Create creative in Meta
        meta_creative_id = await self.meta.create_ad_creative(
            account_id=project.ad_account_id,
            name=f"creative_{generated_name}",
            image_hash=creative.meta_image_hash,
            video_id=creative.meta_video_id,
            headline=headline,
            primary_text=primary_text,
            description=description,
            cta=cta,
            url=final_url,
            url_tags=url_tags,
        )

        # Create ad in Meta
        meta_ad_id = await self.meta.create_ad(
            account_id=project.ad_account_id,
            name=generated_name,
            adset_id=adset.meta_adset_id,
            creative_id=meta_creative_id,
        )

        # Save to DB
        ad = Ad(
            ad_set_id=adset_id,
            name=name,
            generated_name=generated_name,
            creative_ref=creative.base_name,
            headline=headline,
            primary_text=primary_text,
            description=description,
            cta=cta,
            url=final_url,
            url_tags=url_tags,
            meta_ad_id=meta_ad_id,
            meta_creative_id=meta_creative_id,
            status="deployed",
        )
        self.db.add(ad)

        log = DeployLog(
            project_id=project.id, step=DeployStep.ad,
            entity_name=name, meta_id=meta_ad_id, status=DeployStatus.success,
        )
        self.db.add(log)
        await self.db.commit()

        return ad

    async def _get_project(self, project_id: int) -> Project:
        result = await self.db.execute(
            select(Project).where(Project.id == project_id)
        )
        project = result.scalar_one_or_none()
        if not project:
            raise ValueError(f"Project {project_id} not found")
        return project

File: api/lib/services/campaign-manager.ts
/**
 * Post-deploy campaign management: edit budgets, add ad sets/ads to live campaigns.
 */
import { eq } from "drizzle-orm";
import type { Database } from "../db/client.js";
import {
  projects,
  adSets,
  ads,
  creatives,
  deployLogs,
} from "../db/schema.js";
import { MetaApiService } from "./meta-api.js";
import { generateAdSetName, generateAdName } from "./naming.js";
import { buildUrlTags } from "./utm-builder.js";

// ── Inferred row types ──────────────────────────────────────────────────────

type ProjectRow = typeof projects.$inferSelect;
type AdSetRow = typeof adSets.$inferSelect;
type AdRow = typeof ads.$inferSelect;
type CreativeRow = typeof creatives.$inferSelect;

// ── CampaignManager ─────────────────────────────────────────────────────────

export class CampaignManager {
  private db: Database;
  private meta: MetaApiService;

  constructor(db: Database, meta: MetaApiService) {
    this.db = db;
    this.meta = meta;
  }

  /**
   * Update campaign-level budget (CBO).
   */
  async updateBudget(
    projectId: number,
    dailyBudget?: number,
    lifetimeBudget?: number,
  ): Promise<{ status: string; campaign_id: string }> {
    const project = await this.getProject(projectId);
    if (!project.metaCampaignId) {
      throw new Error("Campaign not deployed yet");
    }

    await this.meta.updateBudget(
      project.metaCampaignId,
      dailyBudget,
      lifetimeBudget,
    );

    const updates: Partial<ProjectRow> = {};
    if (dailyBudget != null) {
      updates.dailyBudget = dailyBudget;
    }
    if (lifetimeBudget != null) {
      updates.lifetimeBudget = lifetimeBudget;
    }

    if (Object.keys(updates).length > 0) {
      await this.db
        .update(projects)
        .set(updates)
        .where(eq(projects.id, projectId));
    }

    return { status: "updated", campaign_id: project.metaCampaignId };
  }

  /**
   * Update ad set budget (ABO).
   */
  async updateAdSetBudget(
    adsetId: number,
    dailyBudget: number,
  ): Promise<{ status: string; adset_id: string }> {
    const [adset] = await this.db
      .select()
      .from(adSets)
      .where(eq(adSets.id, adsetId));

    if (!adset || !adset.metaAdsetId) {
      throw new Error("Ad set not found or not deployed");
    }

    await this.meta.updateAdSetBudget(adset.metaAdsetId, dailyBudget);

    await this.db
      .update(adSets)
      .set({ budget: dailyBudget })
      .where(eq(adSets.id, adsetId));

    return { status: "updated", adset_id: adset.metaAdsetId };
  }

  /**
   * Add a new ad set to a deployed campaign.
   */
  async addAdSet(
    projectId: number,
    name: string,
    targetingJson: string,
    optimizationGoal: string = "OFFSITE_CONVERSIONS",
    bidStrategy: string = "LOWEST_COST_WITHOUT_CAP",
    budget?: number,
    placementsJson?: string,
  ): Promise<AdSetRow> {
    const project = await this.getProject(projectId);
    if (!project.metaCampaignId) {
      throw new Error("Campaign not deployed yet");
    }
    if (!project.adAccountId) {
      throw new Error("Project has no ad_account_id set");
    }

    // Count existing ad sets for variant naming
    const existing = await this.db
      .select()
      .from(adSets)
      .where(eq(adSets.projectId, projectId));
    const variantIndex = existing.length;

    const generatedName = generateAdSetName(
      project.name,
      name,
      optimizationGoal,
      variantIndex,
    );

    // Create in Meta
    const targeting: Record<string, unknown> = JSON.parse(targetingJson);
    const placements: Record<string, string[]> | null = placementsJson
      ? JSON.parse(placementsJson)
      : null;

    const metaAdsetId = await this.meta.createAdSet(
      project.adAccountId,
      project.metaCampaignId,
      generatedName,
      optimizationGoal,
      targeting,
      bidStrategy,
      budget,
      placements,
    );

    // Save to DB
    const [newAdSet] = await this.db
      .insert(adSets)
      .values({
        projectId,
        name,
        generatedName,
        targetingJson,
        placementsJson: placementsJson ?? null,
        budget: budget ?? null,
        bidStrategy,
        optimizationGoal,
        metaAdsetId,
        status: "deployed",
      })
      .returning();

    // Log
    await this.db.insert(deployLogs).values({
      projectId,
      step: "adset",
      entityName: name,
      metaId: metaAdsetId,
      status: "success",
    });

    return newAdSet;
  }

  /**
   * Add a new ad to an existing ad set.
   */
  async addAd(
    adsetId: number,
    name: string,
    creativeId: number,
    headline: string = "",
    primaryText: string = "",
    description: string = "",
    cta: string = "SHOP_NOW",
    url?: string,
  ): Promise<AdRow> {
    // Load the ad set
    const [adset] = await this.db
      .select()
      .from(adSets)
      .where(eq(adSets.id, adsetId));

    if (!adset || !adset.metaAdsetId) {
      throw new Error("Ad set not found or not deployed");
    }

    const project = await this.getProject(adset.projectId);
    if (!project.adAccountId) {
      throw new Error("Project has no ad_account_id set");
    }

    // Get creative info
    const [creative] = await this.db
      .select()
      .from(creatives)
      .where(eq(creatives.id, creativeId));

    if (!creative) {
      throw new Error(`Creative ${creativeId} not found`);
    }

    // Ensure creative is uploaded
    if (!creative.metaImageHash && !creative.metaVideoId) {
      if (!creative.localPath) {
        throw new Error(
          `Creative ${creative.originalName} has no local path`,
        );
      }

      if (creative.mediaType === "image") {
        const imageHash = await this.meta.uploadImage(
          project.adAccountId,
          creative.localPath,
        );
        await this.db
          .update(creatives)
          .set({ metaImageHash: imageHash, uploadStatus: "uploaded" })
          .where(eq(creatives.id, creative.id));
        creative.metaImageHash = imageHash;
      } else {
        const videoId = await this.meta.uploadVideo(
          project.adAccountId,
          creative.localPath,
        );
        await this.db
          .update(creatives)
          .set({ metaVideoId: videoId, uploadStatus: "uploaded" })
          .where(eq(creatives.id, creative.id));
        creative.metaVideoId = videoId;
      }
    }

    // Count existing ads for version naming
    const existingAds = await this.db
      .select()
      .from(ads)
      .where(eq(ads.adSetId, adsetId));
    const versionIndex = existingAds.length;

    const generatedName = generateAdName(
      adset.generatedName ?? adset.name,
      creative.baseName ?? name,
      creative.format ?? "",
      versionIndex,
    );

    const urlTags = buildUrlTags();
    const finalUrl = url ?? project.destinationUrl ?? "";

    // Create creative in Meta
    const metaCreativeId = await this.meta.createAdCreative(
      project.adAccountId,
      `creative_${generatedName}`,
      {
        imageHash: creative.metaImageHash ?? undefined,
        videoId: creative.metaVideoId ?? undefined,
        headline,
        primaryText,
        description,
        cta,
        url: finalUrl,
        urlTags,
      },
    );

    // Create ad in Meta
    const metaAdId = await this.meta.createAd(
      project.adAccountId,
      generatedName,
      adset.metaAdsetId,
      metaCreativeId,
    );

    // Save to DB
    const [newAd] = await this.db
      .insert(ads)
      .values({
        adSetId: adsetId,
        name,
        generatedName,
        creativeRef: creative.baseName,
        headline,
        primaryText,
        description,
        cta,
        url: finalUrl,
        urlTags,
        metaAdId,
        metaCreativeId,
        status: "deployed",
      })
      .returning();

    // Log
    await this.db.insert(deployLogs).values({
      projectId: project.id,
      step: "ad",
      entityName: name,
      metaId: metaAdId,
      status: "success",
    });

    return newAd;
  }

  // ── Helpers ───────────────────────────────────────────────────────────────

  private async getProject(projectId: number): Promise<ProjectRow> {
    const [project] = await this.db
      .select()
      .from(projects)
      .where(eq(projects.id, projectId));

    if (!project) {
      throw new Error(`Project ${projectId} not found`);
    }
    return project;
  }
}
